// nuxi/dist/chunks/info.mjs
import os from 'node:os';
import { readFileSync, existsSync } from 'node:fs';
import { createRequire } from 'node:module';
import { t as transform, _ as _createJiti } from '../shared/nuxi.BdnFVL_h.mjs';
import { d as destr } from '../shared/nuxi.D2ykif9w.mjs';
import { s as splitByCase } from '../shared/nuxi.B21HxevS.mjs';
import clipboard from './index6.mjs';
import { d as defineCommand, n as nuxiPkg } from '../shared/nuxi.DSPrE7vg.mjs';
import { g as getPackageManager, a as getPackageManagerVersion } from '../shared/nuxi.B5yd21wq.mjs';
import { f as findup } from '../shared/nuxi.DNA8U7Ya.mjs';
import { c as cwdArgs, a as legacyRootDirArgs, r as resolve } from '../shared/nuxi.BrnifDS2.mjs';
import '../shared/nuxi.E-ZsRS8r.mjs';
import 'node:url';
import 'node:assert';
import 'node:process';
import 'node:path';
import 'node:v8';
import 'node:util';
import 'node:crypto';
import 'node:tty';
import 'node:perf_hooks';
import 'node:vm';
import 'assert';
import 'fs';
import 'module';
import 'os';
import 'path';
import 'process';
import 'tty';
import 'url';
import 'util';
import 'v8';
import './index7.mjs';
import 'node:buffer';
import 'node:child_process';
import '../shared/nuxi.DYqxF_9Q.mjs';
import 'child_process';
import 'node:timers/promises';
import 'stream';

function onError(err) {
  throw err; /* ↓ Check stack trace ↓ */
}

const nativeImport = (id) => import(id);

function createJiti(id, opts = {}) {
  if (!opts.transform) {
    opts = { ...opts, transform };
  }
  return _createJiti(id, opts, {
    onError,
    nativeImport,
    createRequire,
  });
}

const info = defineCommand({
  meta: {
    name: "info",
    description: "Get information about Nuxt project"
  },
  args: {
    ...cwdArgs,
    ...legacyRootDirArgs
  },
  async run(ctx) {
    const cwd = resolve(ctx.args.cwd || ctx.args.rootDir);
    const nuxtConfig = await getNuxtConfig(cwd);
    const { dependencies = {}, devDependencies = {} } = findPackage(cwd);
    const getDepVersion = (name) => getPkg(name, cwd)?.version || dependencies[name] || devDependencies[name];
    function listModules(arr = []) {
      const info = [];
      for (let m of arr) {
        if (Array.isArray(m)) {
          m = m[0];
        }
        const name = normalizeConfigModule(m, cwd);
        if (name) {
          const npmName = name.split("/").splice(0, 2).join("/");
          const v = getDepVersion(npmName);
          info.push("`" + (v ? `${name}@${v}` : name) + "`");
        }
      }
      return info.join(", ");
    }
    const nuxtVersion = getDepVersion("nuxt") || getDepVersion("nuxt-nightly") || getDepVersion("nuxt-edge") || getDepVersion("nuxt3") || "-";
    const isLegacy = nuxtVersion.startsWith("2");
    const builder = !isLegacy ? nuxtConfig.builder || "-" : nuxtConfig.bridge?.vite ? "vite" : nuxtConfig.buildModules?.includes("nuxt-vite") ? "vite" : "webpack";
    let packageManager = getPackageManager(cwd);
    if (packageManager) {
      packageManager += "@" + getPackageManagerVersion(packageManager);
    } else {
      packageManager = "unknown";
    }
    const infoObj = {
      OperatingSystem: os.type(),
      NodeVersion: process.version,
      NuxtVersion: nuxtVersion,
      CLIVersion: nuxiPkg.version,
      NitroVersion: getDepVersion("nitropack"),
      PackageManager: packageManager,
      Builder: typeof builder === "string" ? builder : "custom",
      UserConfig: Object.keys(nuxtConfig).map((key) => "`" + key + "`").join(", "),
      RuntimeModules: listModules(nuxtConfig.modules),
      BuildModules: listModules(nuxtConfig.buildModules || [])
    };
    console.log("Working directory:", cwd);
    let maxLength = 0;
    const entries = Object.entries(infoObj).map(([key, val]) => {
      const label = splitByCase(key).join(" ");
      if (label.length > maxLength) {
        maxLength = label.length;
      }
      return [label, val || "-"];
    });
    let infoStr = "";
    for (const [label, value] of entries) {
      infoStr += "- " + (label + ": ").padEnd(maxLength + 2) + (value.includes("`") ? value : "`" + value + "`") + "\n";
    }
    const copied = await clipboard.write(infoStr).then(() => true).catch(() => false);
    const splitter = "------------------------------";
    console.log(
      `Nuxt project info: ${copied ? "(copied to clipboard)" : ""}

${splitter}
${infoStr}${splitter}
`
    );
    const isNuxt3 = !isLegacy;
    const isBridge = !isNuxt3 && infoObj.BuildModules.includes("bridge");
    const repo = isBridge ? "nuxt/bridge" : "nuxt/nuxt";
    const log = [
      (isNuxt3 || isBridge) && `\u{1F449} Report an issue: https://github.com/${repo}/issues/new?template=bug-report.yml`,
      (isNuxt3 || isBridge) && `\u{1F449} Suggest an improvement: https://github.com/${repo}/discussions/new`,
      `\u{1F449} Read documentation: ${isNuxt3 || isBridge ? "https://nuxt.com" : "https://v2.nuxt.com"}`
    ].filter(Boolean).join("\n");
    console.log("\n" + log + "\n");
  }
});
function normalizeConfigModule(module, rootDir) {
  if (!module) {
    return null;
  }
  if (typeof module === "string") {
    return module.split(rootDir).pop().split("node_modules").pop().replace(/^\//, "");
  }
  if (typeof module === "function") {
    return `${module.name}()`;
  }
  if (Array.isArray(module)) {
    return normalizeConfigModule(module[0], rootDir);
  }
  return null;
}
async function getNuxtConfig(rootDir) {
  try {
    const jiti = createJiti(rootDir, {
      interopDefault: true,
      // allow using `~` and `@` in `nuxt.config`
      alias: {
        "~": rootDir,
        "@": rootDir
      }
    });
    globalThis.defineNuxtConfig = (c) => c;
    const result = await jiti.import("./nuxt.config");
    delete globalThis.defineNuxtConfig;
    return result;
  } catch {
    return {};
  }
}
function getPkg(name, rootDir) {
  let pkgPath = resolve(rootDir, "node_modules", name, "package.json");
  const _require = createRequire(rootDir);
  try {
    pkgPath = _require.resolve(name + "/package.json");
  } catch {
  }
  return readJSONSync(pkgPath);
}
function findPackage(rootDir) {
  return findup(rootDir, (dir) => {
    const p = resolve(dir, "package.json");
    if (existsSync(p)) {
      return readJSONSync(p);
    }
  }) || {};
}
function readJSONSync(filePath) {
  try {
    return destr(readFileSync(filePath, "utf-8"));
  } catch {
    return null;
  }
}

export { info as default };
